library/common/main_interface.h

```diff
@@ -1,12 +1,141 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef int envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP request. Valid only for the duration of the request and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_request_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_request_t request;
+} envoy_request_pair;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_stream_t stream;
+} envoy_stream_pair;
+
+typedef struct {
+  char* name;
+  char* value;
+} envoy_header;
+
+typedef struct {
+  size_t length;
+  envoy_header* headers;
+} envoy_headers;
+
+typedef struct {
+  size_t length;
+  uint8_t* bytes;
+} envoy_data;
+
+const envoy_data envoy_nodata = {0, NULL};
+
+typedef struct {
+  envoy_error_code_t error_code;
+  size_t length;
+  char* string;
+} envoy_error;
+
+#ifdef __cplusplus
+extern "C" { // function pointers
+#endif
+
+typedef void (*on_headers)(envoy_headers headers, bool end_stream);
+typedef void (*on_data)(envoy_data data, bool end_stream);
+typedef void (*on_trailers)(envoy_headers headers);
+typedef void (*on_error)(envoy_error error);
+
+#ifdef __cplusplus
+} // function pointers
+#endif
+
+typedef struct {
+  on_headers h;
+  on_data d;
+  on_trailers t;
+  on_error e;
+} envoy_observer;
+
+#ifdef __cplusplus
+extern "C" { // functions
+#endif
+
+/**
+ * Submit a request.
+ */
+envoy_request_pair submit_request(envoy_headers headers, envoy_data data, envoy_observer observer);
```

Comment:
Should this be `send_request` to match up with the other `send*` functions?

Reply:
We could go further and just call this send, and the stream function start, per the convention in Envoy's AsyncClient. Further, given the return types (envoy_request_pair, envoy_stream_pair) I think the names would be clear that way.

Reply:
Looks like unary upstream uses `send`, and streaming uses `start`. Do we intend to differentiate on this level or some other level (platform/shim)? https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/async_client.h#L210-L221

Reply:
(Seems like maybe we should expose both the streaming and unary APIs?)

Reply:
I don't think I understand your last point. Both send(request) and start(stream) will be available to the platform layer. Internally the common(c/c++) layer will use the `send_<headers/data/trailers>` functions to implement send(request). This is very similar to the Envoy implementation: https://github.com/envoyproxy/envoy/blob/master/source/common/http/async_client_impl.cc#L185 where the AsyncRequestImpl derives from the AsyncStreamImpl.